File: core/log/decorator.py
import functools
import json
import logging
from datetime import datetime
from typing import Any, Optional

from fastapi import Request

from core.db import SessionLocal
from core.utils import get_client_ip, get_city_info, generate_id
from core.utils.trace_utils import get_trace_id
from modules.sys.log.service import LogService
from modules.sys.log.models import SysLog as LogModel
from .utils import parse_user_agent, extract_params_json, get_result_json, generate_log_signature

logger = logging.getLogger(__name__)


def _get_request(*args, **kwargs) -> Optional[Request]:
    request = kwargs.get('request')
    if request:
        return request
    for arg in args:
        if isinstance(arg, Request):
            return arg
    return None


def _get_op_user(request: Request) -> Optional[str]:
    """Get the current operator's display name from the active user."""
    try:
        from core.auth import HeiAuthTool
        import asyncio
        user_id = asyncio.run(HeiAuthTool.getLoginIdDefaultNull(request))
        if not user_id:
            return None

        # Use the same session that _save_log already opened
        from modules.sys.user.service import UserService
        from core.db import SessionLocal
        db = SessionLocal()
        try:
            entity = UserService(db).find_by_id(user_id)
            return entity.nickname or entity.account if entity else None
        finally:
            db.close()
    except Exception:
        return None


def sys_log(name: str = "未命名"):
    """
    Decorator to record operation log on route handlers.
    Logs both success (OPERATE) and exception (EXCEPTION) outcomes.

    Usage:
        @router.get("/api/v1/sys/xxx")
        @SysLog("日志名称")
        @HeiCheckPermission("sys:xxx:page")
        async def handler(request: Request, ...):
            ...
    """
    def decorator(func):
        @functools.wraps(func)
        async def wrapper(*args, **kwargs):
            request = _get_request(*args, **kwargs)
            if not request:
                return await func(*args, **kwargs)

            start_time = datetime.now()
            params_json = extract_params_json(request, kwargs)

            try:
                result = await func(*args, **kwargs)
                _save_log(
                    request=request,
                    func=func,
                    name=name,
                    category="OPERATE",
                    exe_status="SUCCESS",
                    exe_message=None,
                    params_json=params_json,
                    result_json=get_result_json(result),
                    start_time=start_time,
                )
                return result
            except Exception as e:
                _save_log(
                    request=request,
                    func=func,
                    name=name,
                    category="EXCEPTION",
                    exe_status="FAIL",
                    exe_message=str(e)[:2000],
                    params_json=params_json,
                    result_json=None,
                    start_time=start_time,
                )
                raise
        return wrapper
    return decorator


def _save_log(request: Request, func, name: str, category: str,
              exe_status: str, exe_message: Optional[str],
              params_json: str, result_json: Optional[str],
              start_time: datetime) -> None:
    """Persist the log entry to database."""
    db = SessionLocal()
    try:
        now = datetime.now()
        user_agent = request.headers.get("user-agent", "")
        browser, os_name = parse_user_agent(user_agent)
        op_user = _get_op_user(request)

        op_ip = get_client_ip(request)
        entry = LogModel(
            id=generate_id(),
            category=category,
            name=name,
            exe_status=exe_status,
            exe_message=exe_message,
            trace_id=get_trace_id(),
            op_ip=op_ip,
            op_address=get_city_info(op_ip),
            op_browser=browser,
            op_os=os_name,
            class_name=func.__module__,
            method_name=func.__qualname__,
            req_method=request.method,
            req_url=str(request.url),
            param_json=params_json,
            result_json=result_json,
            op_time=start_time,
            op_user=op_user,
            sign_data="",
        )

        sign_data = generate_log_signature({
            "category": category,
            "name": name,
            "exe_status": exe_status,
            "op_ip": entry.op_ip,
            "op_time": start_time.isoformat(),
            "op_user": op_user,
            "trace_id": entry.trace_id,
        })
        entry.sign_data = sign_data

        service = LogService(db)
        service.dao.insert(entry)
        db.commit()
    except Exception as e:
        logger.warning(f"Failed to save operation log: {e}")
        db.rollback()
    finally:
        db.close()

File: core/log/decorator.go
package log

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"hei-gin/core/auth"
	"hei-gin/core/db"
	"hei-gin/core/result"
	"hei-gin/core/utils"
)

// parseUserAgent extracts browser and OS from a User-Agent string.
func parseUserAgent(ua string) (browser, os string) {
	ua = strings.ToLower(ua)
	switch {
	case strings.Contains(ua, "edg"):
		browser = "Edge"
	case strings.Contains(ua, "chrome"):
		browser = "Chrome"
	case strings.Contains(ua, "safari"):
		browser = "Safari"
	case strings.Contains(ua, "firefox"):
		browser = "Firefox"
	case strings.Contains(ua, "opera") || strings.Contains(ua, "opr"):
		browser = "Opera"
	case strings.Contains(ua, "msie") || strings.Contains(ua, "trident"):
		browser = "IE"
	default:
		browser = "Unknown"
	}

	switch {
	case strings.Contains(ua, "windows"):
		os = "Windows"
	case strings.Contains(ua, "mac os") || strings.Contains(ua, "macintosh"):
		os = "macOS"
	case strings.Contains(ua, "linux"):
		os = "Linux"
	case strings.Contains(ua, "android"):
		os = "Android"
	case strings.Contains(ua, "iphone") || strings.Contains(ua, "ipad"):
		os = "iOS"
	default:
		os = "Unknown"
	}
	return
}

type bodyLogWriter struct {
	gin.ResponseWriter
	body *bytes.Buffer
}

func (w *bodyLogWriter) Write(b []byte) (int, error) {
	w.body.Write(b)
	return w.ResponseWriter.Write(b)
}

// SysLog returns a Gin middleware that records an operation log entry.
func SysLog(name string) gin.HandlerFunc {
	return func(c *gin.Context) {
		if db.Client == nil {
			c.Next()
			return
		}

		// Capture request body
		var reqBody []byte
		if c.Request.Body != nil {
			reqBody, _ = io.ReadAll(c.Request.Body)
			c.Request.Body = io.NopCloser(bytes.NewBuffer(reqBody))
		}

		// Wrap response writer
		blw := &bodyLogWriter{body: &bytes.Buffer{}, ResponseWriter: c.Writer}
		c.Writer = blw

		start := time.Now()

		c.Next()

		// Determine status from response code
		category := "OPERATE"
		exeStatus := "SUCCESS"
		exeMessage := ""

		if c.Writer.Status() >= 400 || len(c.Errors) > 0 {
			exeStatus = "FAIL"
			if len(c.Errors) > 0 {
				exeMessage = c.Errors.Last().Err.Error()
			}
		}

		// Parse response to check business code
		respBody := blw.body.Bytes()
		if len(respBody) > 0 {
			var respMap map[string]interface{}
			if json.Unmarshal(respBody, &respMap) == nil {
				if code, ok := respMap["code"].(float64); ok && code >= 400 {
					exeStatus = "FAIL"
					if msg, ok := respMap["message"].(string); ok {
						exeMessage = msg
					}
				}
			}
		}

		// Get current user
		loginType := auth.DetectLoginType(c)
		loginID := ""
		if loginType == "CONSUMER" {
			loginID = auth.ClientAuthTool.GetLoginID(c)
		} else {
			loginID = auth.AuthTool.GetLoginID(c)
		}

		opTime := int(time.Since(start).Milliseconds())
		traceID := result.GetTraceID(c)
		ip := utils.GetClientIP(c)

		// Parse User-Agent
		userAgent := c.GetHeader("User-Agent")
		browser, osName := parseUserAgent(userAgent)

		// Get handler info
		handlerName := c.HandlerName()

		saveLog(c, name, category, exeStatus, exeMessage, loginID, traceID, ip, userAgent,
			browser, osName, handlerName, c.Request.Method, c.Request.URL.String(),
			string(reqBody), string(respBody), opTime)
	}
}

func saveLog(c *gin.Context, name, category, exeStatus, exeMessage, loginID, traceID, ip, userAgent,
	browser, osName, handlerName, reqMethod, reqURL, paramJSON, resultJSON string, opTime int) {

	// Build signature
	signInput := fmt.Sprintf("%s|%s|%s|%s|%s|%s", category, name, exeStatus, ip, time.Now().Format(time.RFC3339), traceID)
	signData := fmt.Sprintf("%x", sha256.Sum256([]byte(signInput)))

	ctx := context.Background()
	now := time.Now()

	_, err := db.Client.SysLog.Create().
		SetID(utils.NextID()).
		SetCategory(category).
		SetName(name).
		SetExeStatus(exeStatus).
		SetExeMessage(exeMessage).
		SetTraceID(traceID).
		SetOpIP(ip).
		SetOpAddress("").
		SetOpBrowser(browser).
		SetOpOs(osName).
		SetClassName(handlerName).
		SetMethodName(c.HandlerName()).
		SetReqMethod(reqMethod).
		SetReqURL(reqURL).
		SetParamJSON(paramJSON).
		SetResultJSON(resultJSON).
		SetOpTime(opTime).
		SetOpUser(loginID).
		SetSignData(signData).
		SetCreatedAt(now).
		SetCreatedBy(loginID).
		Save(ctx)
	if err != nil {
		log.Printf("[SysLog] insert error: %v", err)
	}
}
